### PR TITLE
Added additional filtering to the consumer status report

### PR DESCRIPTION
**Pull request by Ceiu**
Gutterball's consumer status report can now be filtered on (subscription) sku, subscription name or product name. The filtering works much like those provided by the status trend report in that only one is allowed in a given request.

These filter additions required a minor(-ish) change to the object model. Specifically, the ComplianceStatus class required some updates to some of its ElementCollection fields and the JSON converters it used.

Finally, the model changes I opted to use required the addition of an ID field in all three gb_*compprod_snap tables. As such, this commit includes a liquibase update for the tables, and a migration script to populate the ID columns.

The generic filtering API was not added, as it would best fit in common and would be a significant change in itself.

### CODE REVIEW

File: gutterball/src/main/java/org/candlepin/gutterball/liquibase/ComplianceStatusProductId.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        this.logger = logger;
+
+        this.preparedStatements = new HashMap<String, PreparedStatement>();
+    }
+
+    /**
+     *
+     *
+     *
+     */
```
Comment (mstead) [inline gutterball/src/main/java/org/candlepin/gutterball/liquibase/ComplianceStatusProductId.java]: Kill the empty comments or fill it out. Up to yourself.

File: gutterball/src/main/java/org/candlepin/gutterball/liquibase/ComplianceStatusProductId.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * @return
+     *  a 32-character UUID
+     */
+    protected String generateUUID() {
+        return UUID.randomUUID().toString().replace("-", "");
+    }
+
+
+
+
```
Comment (mstead) [inline gutterball/src/main/java/org/candlepin/gutterball/liquibase/ComplianceStatusProductId.java]: Being picky, but wouldn't pass checkstyle with this whitespace  in here.

File: gutterball/src/main/java/org/candlepin/gutterball/report/ConsumerStatusReport.java [hunk 3]
```diff
@@ -151,14 +173,60 @@ protected ReportResult execute(MultivaluedMap<String, String> queryParams, PageR
...
+                consumerIds,
+                ownerFilters,
+                statusFilters,
+                null,
+                null,
+                null,
+                attributeFilters,
+                pageRequest
+            );
+        }
```
Comment (mstead) [inline gutterball/src/main/java/org/candlepin/gutterball/report/ConsumerStatusReport.java]: Would it make sense to create specific methods in ComplianceSnapshotCurator for the calls to getSnapshotIterator above? Would probably be easier to understand what the method is doing. Looks like passing 'null' for certain parameters has a reason... perhaps we can make that more aparent.

File: gutterball/src/main/resources/logback.xml [hunk 4]
```diff
@@ -9,8 +9,8 @@
         </encoder>
     </appender>
 
-    <logger name="org.candlepin.gutterball" level="INFO"/>
-    <logger name="org.candlepin.common" level="INFO"/>
+    <logger name="org.candlepin.gutterball" level="DEBUG"/>
+    <logger name="org.candlepin.common" level="DEBUG"/>
```
Comment (mstead) [inline gutterball/src/main/resources/logback.xml]: Should probably bump this back to INFO?

File: gutterball/src/main/java/org/candlepin/gutterball/report/ConsumerStatusReport.java (hunk 3 above)
Comment (Ceiu) [inline gutterball/src/main/java/org/candlepin/gutterball/report/ConsumerStatusReport.java]: Potentially. This is kind of a stylistic thing derived from the status trend report where the curator/report doesn't actually care about which filters are provided, but combining some of them doesn't make sense logically, so they were disabled at the API level.

File: gutterball/src/test/resources/logback-test.xml [hunk 5]
```diff
@@ -10,7 +10,13 @@
...
   <logger name="liquibase" level="ERROR"/>
 
-  <logger name="org.candlepin.gutterball" level="INFO"/>
+<!--
+  <logger name="org.hibernate.SQL" level="DEBUG"/>
+  <logger name="org.hibernate.type" level="TRACE"/>
+-->
+
+
+  <logger name="org.candlepin.gutterball" level="DEBUG"/>
```
Comment (mstead) [inline gutterball/src/test/resources/logback-test.xml]: I'm torn here... Should this be set to INFO?

File: gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    <changeSet id="20150602102655-7" author="crog">
+        <addPrimaryKey
+            columnNames="id"
+            constraintName="gb_noncompprod_snap_pk"
+            tableName="gb_noncompprod_snap"
+        />
+    </changeSet>
+</databaseChangeLog>
+<!-- vim: set expandtab sts=4 sw=4 ai: -->
```
Comment (mstead) [inline gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml]: Might be worth adding indexes on the columns that we are filtering on.

Comment (Ceiu) [inline gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml]: Primary keys themselves function as indexes, and there are already pre-existing indexes on the other two columns.

File: gutterball/src/test/resources/logback-test.xml (hunk 5 above)
Comment (Ceiu) [inline gutterball/src/test/resources/logback-test.xml]: In test, probably not unless it's generating a lot of noise. I don't really see it hurting anything either way, so I'm okay with whichever.

File: gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml (hunk 6 above)
Comment (mstead) [inline gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml]: ACK.

Comment (mstead) [inline gutterball/src/main/resources/db/changelog/2015-06-02-10-26-compliance-status-products.xml]: > Primary keys themselves function as indexes, and there are already pre-existing indexes on the other two columns.

ACK

**Comment by mstead** [comment]
:+1:  I'll put this feature through its paces today.

**Pull request merged**